### DISCUSSION TITLE
Call to Linq Count() on SqlSet returns an exception ConnectionString not initialized

### DISCUSSION THREAD

**Discussion by ghost** [maxtoroq/DbExtensions#22] [Q&A] [answered]
First of all thanks for this library, it's very powerful and easy to use :+1: 

When I call LinQ Count() method on a SqlSet result, I get an exception "Connectionstring not initialized" :

```
        Tracing.Start("[DALBase.RunQuery] Executing the Query: {0}", query.ToString());

        var result = Enumerable.Empty<TResult>();

        using (var conn = Database.CreateConnection(this.ConnectionString))
        {
            result = conn.Map<TResult>(query);
        }

        Tracing.Stop("[DALBase.RunQuery] Query executed. Number of results: {0}.", result.Count());

        return result;
```

It crashes on the ligne starting with "Tracing.Stop" and returns a System.InvalidOperationException. Here is the stack trace :

```
System.Data.dll!System.Data.SqlClient.SqlConnection.PermissionDemand()  Inconnu
System.Data.dll!System.Data.SqlClient.SqlConnectionFactory.PermissionDemand(System.Data.Common.DbConnection outerConnection)    Inconnu
System.Data.dll!System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(System.Data.Common.DbConnection outerConnection, System.Data.ProviderBase.DbConnectionFactory connectionFactory, System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry, System.Data.Common.DbConnectionOptions userOptions)  Inconnu
System.Data.dll!System.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(System.Data.Common.DbConnection outerConnection, System.Data.ProviderBase.DbConnectionFactory connectionFactory, System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry, System.Data.Common.DbConnectionOptions userOptions)    Inconnu
System.Data.dll!System.Data.SqlClient.SqlConnection.TryOpenInner(System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry)  Inconnu
System.Data.dll!System.Data.SqlClient.SqlConnection.TryOpen(System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry)   Inconnu
System.Data.dll!System.Data.SqlClient.SqlConnection.Open()  Inconnu
```

>   EliasSolutions.Data.DbExtensions.dll!EliasSolutions.Data.DbExtensions.MappingEnumerable<EliasSolutions.DataTransfertObjects.System.LocalizationCategoryDTO>.Enumerator.MoveNext() Ligne 1342    C#
>     System.Core.dll!System.Linq.Enumerable.Count<EliasSolutions.DataTransfertObjects.System.LocalizationCategoryDTO>(System.Collections.Generic.IEnumerable<EliasSolutions.DataTransfertObjects.System.LocalizationCategoryDTO> source) Inconnu

The error occurs in class Enumerator, method MoveNext() on the connection open:
if (prevStateWasClosed)
                  command.Connection.Open();

Any idea what's wrong?

**Comment by maxtoroq** [comment]
Are you sure the connection string is correct?

**Comment by ghost** [comment]
Yes it is correct: Server=DOM\SRVDB;Database=DDEF_DB;Trusted_Connection=True;

During the call to System.Core.dll!System.Linq.Enumerable.Count(System.Collections.Generic.IEnumerable source), when the debugger enters Enumerator.MoveNext(), the command.Connection property is set, but the command.Connection.ConnectionString is empty...

Before the call to System.Core.dll!System.Linq.Enumerable.Count(System.Collections.Generic.IEnumerable source), this property is correctly set.

I don't know what happens in between and why the ConnectionString is empty.

**Comment by maxtoroq** [comment]
You are disposing the connection by using `using(var conn)`.

**Comment by maxtoroq** [comment, marked as answer]
The query is deferred-executed, to query immediately use `ToList()` or `ToArray()`, then you don't have to worry about the connection going out of scope.

**Comment by ghost** [comment]
Ah yes you are totally right! I read that in the documentation but totally forgot to take it into account.

Thanks a lot for your kind help, and sorry about creating an issue regarding this, it was obviously my mistake.

**Comment by maxtoroq** [comment]
No problem, if you have more questions just create an issue.
